internal/controller/controllers/infraenv_controller.go

```diff
@@ -183,10 +184,46 @@ func (r *InfraEnvReconciler) updateInfraEnv(ctx context.Context, log logrus.Fiel
 
 	updateParams.InfraEnvUpdateParams.ImageType = r.Config.ImageType
 
+	existingKargs, err := existingDiscoveryKernelArguments(internalInfraEnv)
+	if err != nil {
+		return nil, err
+	}
+	if !funk.Equal(infraEnv.Spec.DiscoveryKernelArguments, existingKargs) {
+		updateParams.InfraEnvUpdateParams.DiscoveryKernelArguments = discoveryKernelArgumentsParams(infraEnv.Spec.DiscoveryKernelArguments)
+	}
+
 	// UpdateInfraEnvInternal will generate an ISO only if there it was not generated before,
 	return r.Installer.UpdateInfraEnvInternal(ctx, updateParams, nil)
 }
 
+func existingDiscoveryKernelArguments(internalInfraEnv *common.InfraEnv) ([]aiv1beta1.KernelArgument, error) {
+	var args models.KernelArguments
+	if internalInfraEnv.DiscoveryKernelArguments != nil {
+		if err := json.Unmarshal([]byte(swag.StringValue(internalInfraEnv.DiscoveryKernelArguments)), &args); err != nil {
+			return nil, errors.Wrapf(err, "failed to unmarshal discovery kernel arguments of infra-env %s", internalInfraEnv.ID.String())
+		}
+	}
+	var ret []aiv1beta1.KernelArgument
+	for _, arg := range args {
+		ret = append(ret, aiv1beta1.KernelArgument{
+			Operation: arg.Operation,
+			Value:     arg.Value,
+		})
+	}
+	return ret, nil
+}
+
+func discoveryKernelArgumentsParams(kargs []aiv1beta1.KernelArgument) models.KernelArguments {
+	var ret models.KernelArguments
+	for _, arg := range kargs {
+		ret = append(ret, &models.KernelArgument{
+			Operation: arg.Operation,
+			Value:     arg.Value,
+		})
+	}
+	return ret
+}
```

Comment:
The meat of these functions is really transitioning between the two types, right?
The names make it hard to see that.

Maybe something like:
`func kubeKernelArgs(models.KernelArguments) []aiv1beta1KernelArguments`
 and
`func internalKernelArgs([]aiv1beta1.KernelArgument) models.KernelArguments`?

Reply:
I am open about the names ...
But please note that these functions are not the reverse of each other.

Reply:
> But please note that these functions are not the reverse of each other.

I know, but they could be made to be if the unmarshal is moved out of the first one. Since the first function is only called in one place I'm suggesting the parity between the two functions will make them easier to understand and easier to name.

Reply:
OK.  Will change